packages/neptune-extension/src/common/state/notebook/actions.ts

```diff
@@ -1,13 +1,116 @@
+import { ThunkAction, ThunkDispatch } from 'redux-thunk'
+import { AnyAction } from 'redux';
+import {CheckpointDTO, NotebookDTO, NotebookWithNoContentDTO} from 'generated/leaderboard-client/src/models';
+import { leaderboardClient } from "common/api/leaderboard-client";
 
-import { NotebookDTO } from 'generated/leaderboard-client/src/models';
+export const fetchNotebook = (notebookId: string): ThunkAction<Promise<NotebookDTO | void>, {}, {}, NotebookActions> => {
+  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
+    return leaderboardClient.api
+      .getNotebook({ id: notebookId })
+      .then((notebook) => {
+        dispatch(fetchNotebookSuccess(notebook));
+        return notebook;
+      })
+      .catch(() => {
+        dispatch(fetchNotebookFailed());
+        return;
+      })
+  };
+};
 
-export function setNotebook(notebook?: NotebookDTO) {
-  return {
-    type: 'SET_NOTEBOOK',
-    payload: {
-      notebook,
-    },
-  } as const;
+function fetchNotebookSuccess(notebook: NotebookDTO) {
+  return { type: 'NOTEBOOK_FETCH_SUCCESS', payload: {notebook} } as const;
 }
 
-export type NotebookActions = ReturnType<typeof setNotebook>
+function fetchNotebookFailed() {
+  return { type: 'NOTEBOOK_FETCH_FAILED'} as const;
+}
+
+interface CheckpointMetadata {
+  path: string
+  name: string
+  description: string
+}
+
+export const uploadNotebook = (projectId: string, checkpointMeta: CheckpointMetadata, content: any)
```

Comment:
I do not quite understand the intention why the old `uploadNoteboook` and `uploadCheckpoint` utils have been deleted, because here you just rewrote the same with using old then-catch promise handling. Here you could just wrap the old code with redux handling to achieve what you aim to. To me it is preferable that the old utils stay as they clearly demonstrate the flow and function dependencies + use favourable async/await mechanism.

Reply:
Async await vs promise is just a syntax sugar. I dont see any advantage of one over another
If we keep calls in utils, they will function outside of redux flow. We want them to be in redux flow, se we can easily store are errors in store and present them as we please: in Toast or on the Modal.
If we keep the calls in utils, we handle promise/await errors in UplaodModal expanding this component even further. IMHO this component right now is too big and I intended to move as much logic as I can to actions

Reply:
To me it is ok to use redux, as it provides global error handling, just wrap the util code with redux and you are done.